openvm/tests/apc_builder_outputs/single_beqz.txt

```diff
@@ -0,0 +1,27 @@
+// Bus 0 (EXECUTION_BRIDGE):
+mult=-is_valid, args=[from_state__pc_0, reads_aux__1__base__prev_timestamp_0 + reads_aux__1__base__timestamp_lt_aux__lower_decomp__0_0 + 131072 * reads_aux__1__base__timestamp_lt_aux__lower_decomp__1_0]
+mult=is_valid, args=[2 * new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0) + from_state__pc_0 + 2, reads_aux__1__base__prev_timestamp_0 + reads_aux__1__base__timestamp_lt_aux__lower_decomp__0_0 + 131072 * reads_aux__1__base__timestamp_lt_aux__lower_decomp__1_0 + 2]
+
+// Bus 1 (MEMORY):
+mult=is_valid * -1, args=[1, 8, a__0_0, a__1_0, a__2_0, a__3_0, reads_aux__1__base__prev_timestamp_0 + reads_aux__1__base__timestamp_lt_aux__lower_decomp__0_0 + 131072 * reads_aux__1__base__timestamp_lt_aux__lower_decomp__1_0 - (reads_aux__0__base__timestamp_lt_aux__lower_decomp__0_0 + 131072 * reads_aux__0__base__timestamp_lt_aux__lower_decomp__1_0 + 1)]
+mult=is_valid * 1, args=[1, 8, a__0_0, a__1_0, a__2_0, a__3_0, reads_aux__1__base__prev_timestamp_0 + reads_aux__1__base__timestamp_lt_aux__lower_decomp__0_0 + 131072 * reads_aux__1__base__timestamp_lt_aux__lower_decomp__1_0]
+mult=is_valid * -1, args=[1, 0, 0, 0, 0, 0, reads_aux__1__base__prev_timestamp_0]
+mult=is_valid * 1, args=[1, 0, 0, 0, 0, 0, reads_aux__1__base__prev_timestamp_0 + reads_aux__1__base__timestamp_lt_aux__lower_decomp__0_0 + 131072 * reads_aux__1__base__timestamp_lt_aux__lower_decomp__1_0 + 1]
+
+// Bus 2 (PC_LOOKUP):
+mult=is_valid, args=[from_state__pc_0, 4351, 0, 0, 0, 0, 0, 0, 0]
+
+// Bus 3 (VARIABLE_RANGE_CHECKER):
+mult=is_valid * 1, args=[reads_aux__0__base__timestamp_lt_aux__lower_decomp__0_0, 17]
+mult=is_valid * 1, args=[reads_aux__0__base__timestamp_lt_aux__lower_decomp__1_0, 12]
+mult=is_valid * 1, args=[reads_aux__1__base__timestamp_lt_aux__lower_decomp__0_0, 17]
+mult=is_valid * 1, args=[reads_aux__1__base__timestamp_lt_aux__lower_decomp__1_0, 12]
+
+// Algebraic constraints:
+-((1 - new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0)) * (new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0))) = 0
+(1 - new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0)) * a__0_0 = 0
+(1 - new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0)) * a__1_0 = 0
+(1 - new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0)) * a__2_0 = 0
+(1 - new_var2 * (a__0_0 + a__1_0 + a__2_0 + a__3_0)) * a__3_0 = 0
```

Comment:
These three constraints are actually equivalent to the one in line 26 and the reason is quite interesting - we could turn this into an optimization step:
Because of line 21, the first factor in the four constraints is binary. The second factors are byte-constrained because they are read from memory. This means each of the expression in the four constraints are byte range constrained. Now the interesting observation is that `x = 0, y = 0` is equivalent to `x + y = 0` if `x` and `y` are non-negative and the sum does not wrap. So we can turn the set of these four constraints into a single constraint by adding them, and we arrive exactly at the constraint in line 26.